File: android/RevisaoRoomRetrofit/app/src/main/java/com/example/albunsmusicasapp/view/MainActivity.java
package com.example.albunsmusicasapp.view;

import android.os.Bundle;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.SearchView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.albunsmusicasapp.R;
import com.example.albunsmusicasapp.model.Album;
import com.example.albunsmusicasapp.view.adapter.ArtistaAdapter;
import com.example.albunsmusicasapp.viewmodel.ArtistaViewModel;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity{
    private RecyclerView recyclerView;
    private ArtistaAdapter adapter;
    private ArtistaViewModel viewModel;
    private SearchView searchViewPesquisa;
    private ProgressBar loading;
    private List<Album> listaAlbum = new ArrayList<>();
    private String nomeBanda = "Aerosmith";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initViews();

        viewModel.getAlbuns(nomeBanda);

        viewModel.liveDataAlbum.observe(this, albumList -> {
            if (albumList != null && !albumList.isEmpty()) {
                adapter.atualizaLista(albumList);
            } else {
                Toast.makeText(this, "Album não encontrado", Toast.LENGTH_LONG).show();
            }
        });

        viewModel.liveDataLoading.observe(this, loadingBoolean -> {
                    if (loadingBoolean) {
                        loading.setVisibility(View.VISIBLE);
                    } else {
                        loading.setVisibility(View.INVISIBLE);
                    }
                }
        );

        viewModel.liveDataErro.observe(this, error -> {
            Toast.makeText(this, error, Toast.LENGTH_LONG).show();
        });

        searchViewPesquisa.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String novoTexto) {
                nomeBanda = novoTexto;
                adapter.limpaListaAdapter();
                viewModel.getAlbuns(nomeBanda);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String novoTexto) {
                return false;
            }
        });
    }

    private void initViews() {
        recyclerView = findViewById(R.id.recyclerView);
        adapter = new ArtistaAdapter(listaAlbum);
        viewModel = ViewModelProviders.of(this).get(ArtistaViewModel.class);
        searchViewPesquisa = findViewById(R.id.searchView);
        loading = findViewById(R.id.progress_bar);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new GridLayoutManager(this, 2));
    }
}

File: android/KotlinApp/app/src/main/java/com/example/kotlinapp/view/MainActivity.kt
package com.example.kotlinapp.view

import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.kotlinapp.R
import com.example.kotlinapp.model.Filme
import com.example.kotlinapp.view.adapter.FilmeAdapter
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

//        val adapter = FilmeAdapter(getFilmes())
//        recyclerViewFilmes.adapter = adapter
//        recyclerViewFilmes.layoutManager = LinearLayoutManager(this)
//
       entendendoChamadaSegura()
    }


    private fun getFilmes(): MutableList<Filme> {
        val lista = mutableListOf<Filme>()

        lista.add(Filme("JOKER", 0, "TESTE JOKER"))
        lista.add(Filme("LA CASA DE PAPEL", null, null))

        return lista
    }

    private fun entendendoChamadaSegura(){
        val nome =  "joao"
        val outroNome: String? = nome

        Toast.makeText(this, "${outroNome?.length} teste", Toast.LENGTH_LONG).show()

    }
}
